corehq/ex-submodules/couchforms/getters.py

```diff
@@ -54,6 +69,23 @@ def get_instance_and_attachment(request):
     return instance, attachments
 
 
+def _valid_xml_extension(file):
+    return _valid_file_extension(file, PERMITTED_FORM_SUBMISSION_FILE_EXTENSIONS)
+
+
+def _valid_file_extension(file, permitted_extensions):
+    # validate that file has only one extension and that is a permitted one
+    try:
+        file_extension = file.name.split('.', 1)[1]
+    except IndexError:
+        return False
+    return file_extension in permitted_extensions
+
+
+def _valid_attachment_extension(file):
+    return _valid_file_extension(file, SUPPORTED_MEDIA_FILE_EXTENSIONS)
```

Comment:
nit: would be nice if this were moved above `_valid_file_extension` since it is nearly identical to `_valid_xml_extension`

Reply:
Would love to hear your thoughts on,
https://dzone.com/articles/the-stepdown-rule (if you search for "… there are two similar functions in the class?")

As a stepdown rule what i follow is
```
def a():
    b()
    d()

def b():
    c()

def c():
    pass

def d():
    pass
```

Or I can remove the both functions and call `_valid_file_extension` directly if they are not really adding any value.

Reply:
I like and try to use "the stepdown rule" whenever possible. I've also heard it called "The Newspaper Metaphore" in [Chapter 5 of Clean Code](https://www.oreilly.com/library/view/clean-code-a/9780136083238/chapter05.html). That is indeed what motivated this suggestion: `_valid_attachment_extension` calls `_valid_file_extension` and therefore should come before it.